### ISSUE TITLE
Missing of best.pt from resume training

### ISSUE THREAD

**Issue by Chiauwen** [ultralytics/ultralytics#2552]
### Search before asking

- [x] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm training custom data using YOLOv8n with 500 epochs. At the same time, I'm using resume training as a checkpoint to save my best.pt and last.pt after every time they have run for around 20 - 30 epochs. It works kinda fine at the beginning, but I did notice when the total trained epoch has over 100/500, the best.pt will not be generated, however, I can still find the latest last.pt file.

_For example_, whenever I randomly stopped my training at 30/500 (which is before the 100th epoch), and after that I resume it back at the 30th epoch and continue training it for an extra 30 epoch (60/500 epoch), after it finishes at 60th epoch I can still find the latest best.pt and last.pt from runs/detect/train/weights. 
However, after the training has over 100th epochs, and I have resumed training for an extra epoch to reach the final of 500 epochs, I can only find the latest version of last.pt from the weights folder.

I have tried and gone through the same situation 3 times, it will not generate the latest best.pt after the 100th epoch.

I know the question is kinda confusing, but feel free to ask me for any declaration.

I'm using Google Colab for training, and it has limited GPU usage so I have to use resume training when it has reached a certain limit.

**Code I run for resume training:
!yolo task=detect mode=train model="/content/drive/MyDrive/FYP/runs/detect/train/weights/last.pt" data="/content/drive/MyDrive/FYP/data/data.yaml"  epochs=500 save=True resume=true**

Thanks in advance.

**Comment by Chiauwen** [comment]
> Maybe it doesn't improve over 100 epochs ?

The total train, test and valid images have around 7000 images. and I tried the latest best.pt that I trained, the accuracy is not very high. So I think maybe it still can have a further epoch to train.

**Comment by glenn-jocher** [comment]
@Chiauwen hello! Thank you for using YOLOv8 and reaching out to us with your question.

Regarding your observation about the best.pt file not being generated after 100 epochs when using the resume training option, we suggest double-checking the save_best parameter in the train.yaml file. Setting save_best to True should save the best weights on validation dataset if the current mAP is better than the previous best mAP. 

Regarding your question about training beyond 100 epochs, it's possible that your model can continue to improve with further training beyond 100 epochs, especially if you have many images to train on. However, it's important to keep an eye on the training and validation loss to make sure that the model is still learning and not overfitting on the training data. You can also try adjusting the learning rate, using data augmentations and changing other hyperparameters to improve the accuracy of the model if needed. 

I hope this helps! Let us know if you have any other questions or concerns.

**Comment by ercanhocaninyeri** [comment]
I think I have a problem similar to this one. 
When I train a model it generates two .pt files in weights folder of the project: **best.pt** and **last.pt**. After the training finishes, I check the **last change** time of these two files. **best.pt** is always saved 2-3 seconds after **last.pt**. So I assume that the best mAP is achieved at the last step and **best.pt** and **last.pt** are identical files. 
I also check the learning curves (results.png) and the quantitative results (results.csv) to see that the best epoch was much earlier than the last epoch. Therefore, I am sure that these two files should be different. 
I also tried with and without early stopping but the result is the same. 
Here is a screenshot of the model files in my weights folder:
![image](https://github.com/ultralytics/ultralytics/assets/80111184/1092c40a-6b45-4ad3-acfd-ecc32a205d6d)
and the mAP curves for the corresponding training:
![image](https://github.com/ultralytics/ultralytics/assets/80111184/3945cce3-d673-499c-acbd-53310ebeb16e)

It was trained using early stopping with patience=10. The mAP50-95 values for the last 10 epochs (epochs from 20 to 30) are: 
0.69864, 0.66558, 0.33519, 0.62202, 0.66516,  0.68524,  0.67989, 0.69545, 0.68669, 0.69523, 0.68173

Obviously, the model belonging to epoch#20 should be saved as **best.pt** and the model belonging to epoch#30 should be saved as **last.pt**. But they appear be the same file belonging to epoch#30. 
Do you have any comments on this?

I also cannot find the train.yaml file you mentioned in your answer @glenn-jocher. And I cannot see any save_best parameter in any of the .yaml file I use in training process. Could you elaborate on that please?

**Comment by glenn-jocher** [comment]
Hello @ercanhocaninyeri, thank you for reaching out to us regarding this issue. 

Regarding the issue of both best.pt and last.pt being the same file, this behavior is expected in YOLOv8 when the model have not achieved better mAP scores after the best.pt file was saved. As you observed, you can look at the quantitative results to verify whether the best.pt file and last.pt file are different.

Regarding the save_best parameter in train.yaml, that parameter is not a default parameter in YOLOv8. The parameter may have been added in custom training scripts or you can modify the train.yaml file to add the save_best parameter. Setting save_best to True should save the best weights on validation dataset if the current mAP is better than the previous best mAP.

I hope that helps! Feel free to let us know if you have any other questions or concerns.

**Comment by wshi999** [comment]
Hello @glenn-jocher.
I failed find any information about `save_best` in yolov8 doc. And when I try to set it to `True` in the training yaml file or python argument, I was prompted that this argument does not exist.
Could you point me to the doc regarding this or show me an example of how to provide this argument?

**Comment by ExtReMLapin** [comment]
I mean it should be on by default

**Comment by glenn-jocher** [comment]
Hello @wshi999, thank you for bringing this to our attention.

In YOLOv8, the functionality to save the best model based on mAP is not currently available by default. The best.pt and last.pt files are always the same if the model does not achieve a better mAP after the best.pt file was saved. This behavior is intentional and helps prevent unnecessary duplication of model weights files.

If you would like to implement the functionality to save the best model based on mAP yourself, you can modify the training script to track the best mAP value during training and save the corresponding model weights separately.

Let me know if there's anything else I can assist you with!

**Comment by Gnoot01** [comment]
@glenn-jocher Hi, I would like to ask why resuming from a last.pt checkpoint results in no more best.pt checkpoints being created anymore after? there's always only last.pt after resuming but not both last.pt and best.pt like if it didn't have to resume.

**Comment by glenn-jocher** [comment]
Hello @Gnoot01, when resuming training, the `
